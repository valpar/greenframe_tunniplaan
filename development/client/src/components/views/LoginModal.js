import React, { Fragment, useEffect } from "react";
import ReactDOM from "react-dom";
import Input from "../UI/Input/Input";

const Backdrop = (props) => {
  useEffect(() => {
    const themeColorMeta = document.querySelector('meta[name="theme-color"]');

    themeColorMeta.setAttribute("content", "#5C5C5C");

    return () => {
      themeColorMeta.setAttribute("content", "#fff");
    };
  }, []);

  return (
    <div
      data-testid="backdrop"
      className="fixed top-0 left-0 w-full h-screen z-20 bg-neutral-900 opacity-70"
      onClick={props.onClose}
    />
  );
};

const ConfirmModalOverlay = (props) => {
  
  const changeHandler = (input) => {
    
    if (input.name==="email") {
      props.onEmailChange(input.value);
    }
    if (input.name==="password") {
      props.onPasswordChange(input.value);
    }
  }
  

  let arrow =
    "absolute z-40 w-8 h-8 -rotate-45 -right-4 bg-white top-1/2 -translate-y-1/2";
  if (props.topArrow)
    arrow =
      "absolute z-40 w-8 h-8 -rotate-45 -top-4 bg-white left-1/2 -translate-x-1/2";
  if (props.bottomArrow)
    arrow =
      "absolute z-40 w-8 h-8 -rotate-45 -bottom-4 bg-white left-1/2 -translate-x-1/2";

  return (
    <div className="fixed top-1/2 -translate-y-1/2 left-1/2 -translate-x-1/2 lg:translate-x-0 lg:translate-y-0 lg:left-auto lg:top-auto lg:relative w-auto h-auto space-y-2 bg-white p-4 rounded z-30 shadow-lg">
      <div className={`hidden lg:block ${arrow}`}></div>
      <div className="pb-4 text-center">
        <p>{props.modalMessage}</p>
      </div>

      <label>
        E-mail: 
        <Input
          name="email" 
          value={props.email} 
          onChange={changeHandler} 
          readOnly={false}
          className={`p-2 border ${
            props.hasError || props.errorMessage
              ? "border-red-500"
              : "border-borderBlack "
          } rounded-none font-normal focus:outline-none focus:border-black`}
        />
      </label>
      <label>
        Parool: 
        <Input 
          name="password" 
          value={props.password} 
          onChange={changeHandler} 
          type="password" readOnly={false} 
          className={`p-2 border ${
            props.hasError || props.errorMessage
              ? "border-red-500"
              : "border-borderGray "
          } rounded-none font-normal focus:outline-none focus:border-black`}
        />
      </label>


      <div className="flex space-x-4">
        <button
          onClick={props.onDecline}
          type="button"
          className="border border-borderGray py-1 px-2 w-24 hover:bg-borderGray"
        >
          {props.modalButtons ? props.modalButtons[0] : `EI`}
        </button>
        <button
          onClick={props.onConfirm}
          type="button"
          className="border border-borderGray py-1 px-2 w-24 hover:bg-borderGray"
        >
          {props.modalButtons ? props.modalButtons[1] : `JAH`}
        </button>
      </div>
    </div>
  );
};



const portalElement = document.getElementById("overlays");

const LoginModal = (props) => {

  return (
    
    <Fragment>
      {ReactDOM.createPortal(<Backdrop />, portalElement)}
      <ConfirmModalOverlay {...props} /> 
    </Fragment>
    
  );
};

export default LoginModal;
