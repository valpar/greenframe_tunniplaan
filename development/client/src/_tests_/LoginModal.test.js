import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import LoginModal from '../components/views/LoginModal';

beforeAll(() => {
    const meta = document.createElement('meta');
    meta.setAttribute('name', 'theme-color');
    document.head.appendChild(meta);
  });
  
afterAll(() => {
  // eslint-disable-next-line testing-library/no-node-access
  document.head.removeChild(document.querySelector('meta[name="theme-color"]'));
});

// kui seda ei tee (createPortal), siis target "is not DOM element"
jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: (element, target) => element, 
}));

describe('LoginModal', () => {
  const mockOnClose = jest.fn();
  const mockOnEmailChange = jest.fn();
  const mockOnPasswordChange = jest.fn();
  const mockOnDecline = jest.fn();
  const mockOnConfirm = jest.fn();
  it("matches snapshot", () => {
    const tree = renderer.create(
      <LoginModal
        onClose={mockOnClose}
        onEmailChange={mockOnEmailChange}
        onPasswordChange={mockOnPasswordChange}
        onDecline={mockOnDecline}
        onConfirm={mockOnConfirm}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders with Backdrop', () => {
    render(<LoginModal />);
    
    expect(screen.getByTestId('backdrop')).toBeInTheDocument();
  });
});