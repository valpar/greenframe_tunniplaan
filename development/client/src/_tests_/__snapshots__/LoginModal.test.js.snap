// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginModal matches snapshot 1`] = `
Array [
  <div
    className="fixed top-0 left-0 w-full h-screen z-20 bg-neutral-900 opacity-70"
    data-testid="backdrop"
  />,
  <div
    className="fixed top-1/2 -translate-y-1/2 left-1/2 -translate-x-1/2 lg:translate-x-0 lg:translate-y-0 lg:left-auto lg:top-auto lg:relative w-auto h-auto space-y-2 bg-white p-4 rounded z-30 shadow-lg"
  >
    <div
      className="hidden lg:block absolute z-40 w-8 h-8 -rotate-45 -right-4 bg-white top-1/2 -translate-y-1/2"
    />
    <div
      className="pb-4 text-center"
    >
      <p />
    </div>
    <label>
      E-mail:
      <input
        autoComplete="off"
        className=""
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="text"
        value=""
      />
    </label>
    <label>
      Parool:
      <input
        autoComplete="off"
        className=""
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="password"
        value=""
      />
    </label>
    <div
      className="flex space-x-4"
    >
      <button
        className="border border-borderGray py-1 px-2 w-24 hover:bg-borderGray"
        onClick={[MockFunction]}
        type="button"
      >
        EI
      </button>
      <button
        className="border border-borderGray py-1 px-2 w-24 hover:bg-borderGray"
        onClick={[MockFunction]}
        type="button"
      >
        JAH
      </button>
    </div>
  </div>,
]
`;
